crates/noirc_evaluator/src/ssa/opt/mem2reg.rs

```diff
@@ -298,22 +300,85 @@ impl PerFunctionContext {
                 aliases.insert(result);
                 references.aliases.insert(Expression::Other(result), aliases);
             }
+            Instruction::ArrayGet { array, .. } => {
+                let result = function.dfg.instruction_results(instruction)[0];
+                references.mark_value_used(*array, function);
+
+                if function.dfg.value_is_reference(result) {
+                    let array = function.dfg.resolve(*array);
+                    let expression = Expression::ArrayElement(Box::new(Expression::Other(array)));
+
+                    if let Some(aliases) = references.aliases.get_mut(&expression) {
+                        aliases.insert(result);
+                    }
+                }
+            }
+            Instruction::ArraySet { array, value, .. } => {
+                references.mark_value_used(*array, function);
+
+                if function.dfg.value_is_reference(*value) {
+                    let result = function.dfg.instruction_results(instruction)[0];
+                    let array = function.dfg.resolve(*array);
+
+                    let expression = Expression::ArrayElement(Box::new(Expression::Other(array)));
+
+                    if let Some(aliases) = references.aliases.get_mut(&expression) {
+                        aliases.insert(result);
+                    } else if let Some((elements, _)) = function.dfg.get_array_constant(array) {
+                        // TODO: This should be a unification of each alias set
+                        // If any are empty, the whole should be as well.
```

Comment:
Is this handled in the followup PR or should we make an issue?

Reply:
I'll make an issue for it, its not handled in the mem2reg constant folding PR. In general we still need to handle other operations such as the builtin slice functions as well in case these return slices with references in them.

Reply:
https://github.com/noir-lang/noir/issues/2477